### ISSUE TITLE
HTTPS Support

### ISSUE THREAD

**Issue by Moosems** [saviorand/lightbug_http#20]
**Is your feature request related to a problem? Please describe.**
Would be nice to be able to say that a website I made with this is "secure" (the s in HTTPS)

**Describe the solution you'd like**
A way to make HTTPS connections OR HTTP connections

**Describe alternatives you've considered**
Just dealing with the current implementation?

**Comment by saviorand** [comment]
This also came up in #28 , I think this is very important to drive adoption more generally.
Added a `good first issue` label, contributions very welcome!

**Comment by saviorand** [comment]
Actually now that I think about it, it's not a `good first issue`. But still, would welcome any contributions

**Comment by vikramsg** [comment]
Hi,

Thanks for creating this lib. I was curious about HTTP libraries in Mojo and came across this one. While going through this issue I realized that other languages have stuff in their standard library that is  missing in Mojo to enabled HTTPS. 

I tried to figure out how for eg. `starlette` does it and it turns out it `starlette` doesn’t and outsources it to the underlying http server which usually is `uvicorn`. And how does `uvicorn` do it? It outsources the [TLS](https://github.com/encode/uvicorn/blob/c23cd24e6676ee0638d014d7000af1e6e0996bd6/uvicorn/server.py#L134) part to `asyncio` which in turns outsources it to the Python SSL library. 

I don’t have enough Go understanding, but I think FastHTTP also does the same thing and outsources the TLS part to [crypto/TLS](https://github.com/valyala/fasthttp/blob/b06f4e21d918faa84ae0aa12c9e4dc7285b9767e/server.go#L6).

I feel like to get HTTPS support will require the following steps. 

1. Wrap [OpenSSL](https://www.openssl.org/) or create SSL from scratch. 
2. Wrap socket connections with SSL which will handle the TLS handshake process etc. 

Does that sound like the way to go?

**Comment by saviorand** [comment]
@vikramsg you're completely right, this requires spinning up a separate TLS library at this point. I think it makes sense to keep that one in a different repo/project similar to how it's done in Go or Python like you shared.
I've also [asked in Modular Discord](https://discord.com/channels/1087530497313357884/1262445909321453610/1262445909321453610), Jack confirmed that team doesn't plan to add TLS to the stdlib any time soon and they welcome community contributions there. If you're up for the challenge happy to collaborate on the OpenSSL wrapper library, and of course once that's available happy to start working on adding TLS socket wrapper to Lightbug as soon as possible

**Comment by vikramsg** [comment]
@saviorand thanks for the reply. 

Yes, I think it makes sense to do a separate repo for this. Do you have some ideas on the scope of what we want. I wouldn't want to try to support all of OpenSSL, just the bits that are required for HTTPS.

**Comment by saviorand** [comment]
@vikramsg what would be nice to have in a hypothetical TLS library are the utilities for wrapping connections/listeners into their "secure" versions that would conduct handshakes/checks when needed. I imagine for that we'll need things like (modeling off of Lightbug's types, although these don't have to match directly); examples are from Go's `crypto/tls`:
1. A [config struct](https://cs.opensource.google/go/go/+/go1.22.5:src/crypto/tls/common.go;l=527) at a minimum just holding one or more certificates and maybe authorities, although we can use host's root authority set for now. Some parsing primitives e.g. for parsing an X509 keypair, so we can [load it from a file](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/tls.go;drc=a0da9c00aeb51261b9845a46fbc9878870884ab6;l=230)
2. A [listener](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/tls.go;l=77) and a [server](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/tls.go;l=34) with handshake functions for a [client](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/handshake_client.go;drc=81fc6162675ea2543aa30ae35e56769c3d8a39ca;l=173) and a [server](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/handshake_server.go;drc=81fc6162675ea2543aa30ae35e56769c3d8a39ca;l=40)
3. A [connection](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/conn.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=28) that can e.g. hold handshake data and has methods for [read](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/conn.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=1356), [write](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/conn.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=971) and also does things like [encrypt](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/conn.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=479)/[decrypt](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/crypto/tls/conn.go;drc=059a9eedf45f4909db6a24242c106be15fb27193;l=341).

A `crypto` package is in general missing from the Mojo stdlib,  so we cannot call things like `crypto/aes`, `crypto/ecdsa`, `crypto/rsa` or [`crypto/x509`](https://pkg.go.dev/crypto/x509) like done here from within Go's `crypto/tls`. I've searched for community libraries for this but couldn't find anything we could use.. So also a major gap looks like.

Another challenge of course is keeping such a TLS library general/not tied to TCP/HTTP since TLS can potentially be used to secure any kind of transport format, not just in the form of HTTP. Since not much is implemented in Mojo in terms of other protocols in my opinion we can go easy on the universality part for now.

**Comment by saviorand** [comment]
A basic flow to support would be like something described here https://github.com/denji/golang-tls

**Comment by saviorand** [comment]
@vikramsg thought it might be interesting for you, we're collaborating on a Rustls-FFI based TLS implementation with @thatstoasty and @lsh here https://github.com/thatstoasty/mojo-rustls/ , once it's done I could import it as a dependency into Lightbug and try making an HTTPS client/server

**Comment by vikramsg** [comment]
@saviorand thanks for the update. Unfortunately I haven't been able to spend any time on this.

Wish you good luck and looking forward to the implementation.

**Comment by vikramsg** [comment]
Hey @saviorand,

I spent the weekend figuring out how much I could push AI agents which resulted in https://github.com/vikramsg/mossl. 
Its almost purely vibe coded but with a lot of steering and a lot of tests. 

Net result is that with `lightbug_http` I can do HTTPS requests for sites that support TLS1.3 - https://github.com/vikramsg/mossl/blob/main/scripts/https_get.mojo

So, the next question is, is this of interest to you to create a full HTTPS client in `lightbug_http`. If so, we can start discussing further?

**Comment by saviorand** [comment]
Hi @vikramsg !
This looks really cool, but we're currently in the process of removing the client from lightbug in favor of @thatstoasty 's new client library:
https://github.com/thatstoasty/mojo-curl
It's based on curl though, so we probably can't use your code there.

I really appreciate the effort though, and looking at your code, it looks like at least two things can be extracted into their own libraries that could be useful to the community: 1) crypto primitives , and 2) ssl bindings . 

Perhaps the rest of the code could be extracted into its own separate non-curl based client library, decoupled from lightbug. Feel free to borrow existing lightbug client code for this if it's useful. Thanks!

**Comment by vikramsg** [comment]
> Perhaps the rest of the code could be extracted into its own separate non-curl based client library

Ok, for my understanding, there will be no HTTP client in this repo anymore?

> I really appreciate the effort though, and looking at your code, it looks like at least two things can be extracted into their own libraries that could be useful to the community: 1) crypto primitives , and 2) ssl bindings .

> Perhaps the rest of the code could be extracted into its own separate non-curl based client library, decoupled from lightbug. Feel free to borrow existing lightbug client code for this if it's useful

If I understand this right, you are recommending 3 different projects for this to be useful. 

1. Crypto lib
2. SSL bindings (probably something like a TLS1.3 only lib)
3. A HTTP/HTTPS client library that can use the crypto and TLS but since no lightbug http lib anymore, so will need some replication.

In that case, can you tell me more about what will remain in this repo? Socket implementation, header parsers?

Thanks for the help.
